desktop_version/src/Music.cpp

```diff
@@ -92,81 +131,263 @@ class SoundTrack
+        if (looplength == 0) {
+            faudio_buffer.PlayLength = stb_vorbis_get_samples_float_interleaved(vorbis, channels, (float*)decoded_buf_playing, size/sizeof(float));
+        } else {
+            faudio_buffer.PlayLength = std::min(stb_vorbis_get_samples_float_interleaved(vorbis, channels, (float*)decoded_buf_playing, size/sizeof(float)), (loopbegin+looplength)-sample_pos);
+        }
+        faudio_buffer.AudioBytes = size;
+        faudio_buffer.pAudioData = decoded_buf_playing;
+        faudio_buffer.pContext = this;
+        sample_pos += faudio_buffer.PlayLength;
+        if (FAudioSourceVoice_SubmitSourceBuffer(musicVoice, &faudio_buffer, NULL)) {
+            vlog_error("Unable to queue sound buffer");
             return false;
         }
+        Resume();
         return true;
     }
 
     static void Halt()
     {
         /* FAudioVoice_Destroy */
-        Mix_HaltMusic();
+        if (!IsHalted()) {
+            FAudioSourceVoice_FlushSourceBuffers(musicVoice);
+            FAudioVoice_DestroyVoice(musicVoice);
+            musicVoice = NULL;
+        }
     }
 
     static bool IsHalted()
     {
-        /* return musicVoice == NULL; */
-        return Mix_PausedMusic() == 1;
+        return musicVoice == NULL;
     }
 
     static void Pause()
     {
-        /* FAudioSourceVoice_Pause */
-        Mix_PauseMusic();
+        vlog_info("paused musictracks");
+        if (!IsHalted()) FAudioSourceVoice_Stop(musicVoice, 0, FAUDIO_COMMIT_NOW);
     }
 
     static void Resume()
     {
-        /* FAudioSourceVoice_Resume */
-        Mix_ResumeMusic();
+        vlog_info("resumed musictracks");
+        if (!IsHalted()) FAudioSourceVoice_Start(musicVoice, 0, FAUDIO_COMMIT_NOW);
     }
 
     static void SetVolume(int musicVolume)
     {
-        /* FAudioSourceVoice_SetVolume */
-        Mix_VolumeMusic(musicVolume);
+        float adj_vol = ((float)musicVolume/VVV_MAX_VOLUME)*2;
+        if (!IsHalted()) FAudioVoice_SetVolume(musicVoice, adj_vol, FAUDIO_COMMIT_NOW);
     }
 
 private:
-    Mix_Music *m_music;
-};
+    stb_vorbis* vorbis = NULL;
+    int channels;
+    Uint32 size;
+    int loopbegin;
+    int looplength;
+    int sample_pos; //stb_vorbis offset not yet functional on pulldata API. TODO Replace when fixed
+
+    FAudioVoiceCallback callbacks;
+    FAudioWaveFormatEx format;
+
+    Uint8* decoded_buf_playing = NULL;
+    Uint8* decoded_buf_reserve = NULL;
+    Uint8* read_buf = NULL;
+    bool shouldloop;
+
+    static FAudioSourceVoice* musicVoice;
+
+    static void refillReserve(FAudioVoiceCallback* callback, void* ctx) {
+        MusicTrack* t = (MusicTrack*)ctx;
+        FAudioBuffer faudio_buffer;
+        SDL_memset(&faudio_buffer, 0, sizeof(FAudioBuffer));
+        if (t->looplength == 0) {
+            faudio_buffer.PlayLength = stb_vorbis_get_samples_float_interleaved(t->vorbis, t->channels, (float*)t->decoded_buf_reserve, t->size/sizeof(float));
+        } else {
```

Comment:
Closing brace needs to be on a separate line. Opening brace needs to be on the next line.